### ISSUE TITLE
Moving curator-spawned units under __SERVER__

### ISSUE THREAD

**Issue by freghar** [CntoDev/cnto-mmt-repo#89]
Zeus-spawned units remain under AI control of the player's client, see
http://carpenoctem.co/forums/m/26081621/viewthread/26113770-object-locality-game-performance

The actual fix for this is simple, `(group _this) setGroupOwner 2;` wherever the unit is local. Ideally, this would be hooked to [CuratorObjectPlaced](https://community.bistudio.com/wiki/Arma_3:_Event_Handlers#CuratorObjectPlaced), so that any spawned unit is automatically transferred under \__SERVER__ (ID 2).

This is reasonably easy for the editor-placed Curator object, but is not so clear how to hook it up to the MCC curator instance (which is a separate one). We could iterate over allCurators on mission start, but I'm not sure if the MCC curator is not created/destroyed when player logs into / out of MCC admin interface.

@xand3rr , @milivojm : Any ideas how the MCC curator instance works?

**Comment by milivojm** [comment]
Hey, please see what @HellfireLx has written in:

http://www.carpenoctem.co/forums/m/26081621/viewthread/24764187-making-nonzeus-units-controllable-by-zeus

In essence, there is a variable `MCC_curator` which represents MCC Zeus. Also, he suggests a better approach in dealing with curators. If only he was not so busy, right Hellfire? :)

**Comment by freghar** [comment]
@milivojm Thanks for the pointer, however the problem with registering the CuratorObjectPlaced EH is that we need to do it only once, we can't just hook CuratorObjectRegistered (curator interface opened) or something similar, because it would stack multiple CuratorObjectPlaced EHs, one per each interface open. We also cannot do it "on-demand" using getAssignedCuratorLogic, for the same reason.

Each curator instance needs to have only one registration of CuratorObjectPlaced per its existence, which probably means one per mission start (unless MCC destroys the instance when player logs out, which it doesn't seem to).

Luckily, it seems that it's going to be easier than I thought, MCC_curator is simply created in init.sqf (in MCC pbo) and this EH is already there:

    //Curator custom menus
    //Add eventHandler when editing object
    MCC_curator addEventHandler ["CuratorObjectDoubleClicked", {_this spawn MCC_fnc_curatorInitLine}];

Now the problem is figuring out how MCC deals with this curator object, which is not easy. I'm not sure if addon's init.sqf is called during game init (my guess based on other work) or mission init. The MCC_curator instance also doesn't seem to be publicVariable, so each client has its own and if it's initialized on game start, it would be persistent across missions, right?
Except MCC also does `BIS_fnc_activateAddons` in `mcc/fnc/general/fn_activateAddons.sqf` explicitly, so I'm not sure that's true.

If we hook the CuratorObjectPlaced EH in our mission init.sqf (guarded by `!isServer`), won't it get added multiple times if the player disconnects and re-joins?

**Comment by freghar** [comment]
Okay, it seems that when I `deleteVehicle` the MCC curator instance (the only one in `allCurators`),
 * `MCC_curator` is re-initialized on every editor Play in Singleplayer action, so it's not game-global
 * it also gets reinitialized on `#restart` of a mission, even on dedicated
 * it is **not** reinitialized on JIP, but equally it is not empty when I JIP without deleting it first

This leaves me slightly confused, because:
 * `MCC_curator` is nowhere defined as publicVariable
 * any (even global) variables are always removed on leave/join JIP cycle

So `MCC_curator` is *somehow* synchronized across clients/server, but I'm currently not sure how. Wiki says `deleteVehicle` has Global effect, so if `MCC_curator` was the same on my client as on the server, there might be nothing to sychronize, which would explain empty `allCurators` on JIP, after deleting `MCC_curator`.

**Comment by freghar** [comment]
So `MCC_curator` is a publicVariable, it's just too obfuscated to understand at a first glance:

        //CURATOR
        _dummy = MCC_dummyLogicGroup createunit ["ModuleCurator_F", [0, 90, 90],[],0.5,"NONE"]; //Logic Server
        _name = "MCC_curator";
        _dummy setvariable ["text","MCC_curator"];
        _dummy setvariable ["mccIgnore",true];
        _dummy setvariable ["Addons",2,true];
        _dummy setvariable ["vehicleinit","_this setvariable ['Addons',2,true];"];
        call compile (_name + " = _dummy");
        publicVariable _name;

The key part being `call compile (_name + " = _dummy");` which translates to `call { MCC_curator = _dummy; }`, followed by `publicVariable "MCC_curator"`.

Okay, so I'm not crazy.

The entire thing is also in `if ( isServer ) then`, which makes sense - you don't want multiple clients declaring the same publicVariable. Based on this, we should be able to just add our EH in mission init.sqf wrapped in `isServer`, to not execute it on every client.

**Comment by freghar** [comment]
Something like this should do it (in `init.sqf`), untested:

    // CNTO - move ownership of Zeus-spawned units under __SERVER__
    if (isServer && !isNull MCC_curator) then {
        MCC_curator addEventHandler ["CuratorObjectPlaced", {
            params ["_curator", "_unit"];
            private _group = group _unit;
            if (groupOwner _group != 2) then {
                _group setGroupOwner 2;
            };
        }];
    };

@CNTODachi , which location in this (cnto-mmt-repo) repo should I use? .. There seems to be .pbo in `Mission Making Team\Missions\Templates\Official\Development`, but that doesn't match the Template @milivojm  linked in issue #88 .

**Comment by CNTODachi** [comment]
Hey @freghar 

The Template in `Mission Making Team\Mission\Templates\Official\Development` is an updated version of @milivojm template in which I added the ACE3 Module for Fast Roping already attached to all Compatible Helicopters within our repo and marked on the map with a Marker saying the following `Optional ACE3 Fast Roping Module & Compatible Heli's`.

Still working on that template as we are having several issues with LOP at the minute, will update you after further testing. But feel free to use that template to create an alternate version in order to test this functionality although be aware testing this will probably be low priority right now due the fact if not implemented it will not be game breaking.

Regards,
Dachi

**Comment by freghar** [comment]
@CNTODachi The pbo in `Development` doesn't have any playable slots in the `mission.sqm`, though. This issue is reasonably high priority for me, as it would allow us to use server-side ASR AI settings. The code I posted won't work, though, as the EH is executed on the client, the spawned units are local to the client, but the `setGroupOwner` can be run only by the server, so I'll need some more `BIS_fnc_MP` magic.

**Comment by milivojm** [comment]
@freghar apparently they've introduced remoteExec instead.

**Comment by freghar** [comment]
@milivojm I know, that's what I meant. I keep referring to the family of remote exec functions as `BIS_fnc_MP` for simplicity, since (as of 1.52 or so) they share the same code - all need CfgRemoteExec permissions and have the same performance (unlike older versions). :)

Anyway, I have a working version now, just tuning it for which units to apply it on.
